### PR TITLE
Add quoted information to CSV::FieldInfo #252

### PR DESCRIPTION
**Pull request by heronshoes**
This is a proposal to fix the issue #252;
- Add a new member `:quoted?` to `FieldInfo` .
- ~~Introduce an array of booleans `@quoted_fields` in Parser to collect original column value was quoted?.~~
- Check in each of `parse_no_quote`, `parse_quotable_loose` and `parse_quotable_robust`.
  - Use local variable `quoted_fields` to collect booleans in each method and pass it through `#emit_row`.
- Add test of custom converter for preserving quoted value.

### CODE REVIEW

File: lib/csv/fields_converter.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def empty?
       @converters.empty?
     end
 
-    def convert(fields, headers, lineno)
+    def convert(fields, headers, lineno, quoted_fields = nil)
```
Comment (kou) [inline lib/csv/fields_converter.rb:47]: Can we make `quoted_fields` mandatory argument? If we use `nil` here, `FieldInfo#quoted?` returns `nil` not boolean.

File: lib/csv/parser.rb [hunk 2]
```diff
@@ -928,9 +928,11 @@ def parse_no_quote(&block)
         if line.empty?
           next if @skip_blanks
           row = []
+          @quoted_fields = []
```
Comment (kou) [inline lib/csv/parser.rb:931]: How about passing `quoted_fields` to `#emit_row` and make this a local variable?

File: lib/csv/fields_converter.rb (hunk 1 above)
Comment (heronshoes) [reply lib/csv/fields_converter.rb:47]: Thanks. I thought it is not required when converting headers.
But it is better to explicitly pass an argument for them.

File: lib/csv/parser.rb (hunk 2 above)
Comment (heronshoes) [reply lib/csv/parser.rb:931]: Thanks.
This is simpler and localized !

File: lib/csv/fields_converter.rb (hunk 1 above)
Comment (kou) [reply lib/csv/fields_converter.rb]: Headers may be parsed from string. So quoted information for headers may be useful.

Comment (heronshoes) [reply lib/csv/fields_converter.rb]: I understood, thank you. So currently no information is passed in `#adjust_headers`  ....

Comment (heronshoes) [reply lib/csv/fields_converter.rb]: I added support for quoted header converters.

**Comment by heronshoes** [comment]
I rebased feat/issue-252 to the latest HEAD.

File: lib/csv/parser.rb [hunk 3]
```diff
@@ -783,8 +783,9 @@ def parse_headers(row)
                      quote_char: @quote_character)
     end
 
-    def adjust_headers(headers)
-      adjusted_headers = @header_fields_converter.convert(headers, nil, @lineno)
+    def adjust_headers(headers, quoted_fields = nil)
```
Comment (kou) [inline lib/csv/parser.rb:786]: Can we always require `quoted_fields` like `emit_raw`?

```suggestion
    def adjust_headers(headers, quoted_fields)
```

File: lib/csv/parser.rb [hunk 4]
```diff
@@ -1021,18 +1030,19 @@ def parse_quotable_robust(&block)
           row << value
```
Comment (kou) [inline lib/csv/parser.rb:1030]: Do we need add an element to `quoted_fields` here?

```suggestion
          row << value
          quoted_fields << @quoted_column_value
```

Could you also add a test that uses this method? `CSV.parse(..., strip: true)` will use this method.

File: lib/csv/parser.rb [hunk 5]
```diff
@@ -1021,18 +1030,19 @@ def parse_quotable_robust(&block)
...
         elsif parse_row_end
           if row.empty? and value.nil?
-            emit_row([], &block) unless @skip_blanks
+            emit_row([], [], &block) unless @skip_blanks
           else
             row << value
```
Comment (kou) [inline lib/csv/parser.rb:1035]: Do we need add an element to `quoted_field`s here?

```suggestion
            row << value
            quoted_fields << @quoted_column_value
```

File: lib/csv/parser.rb [hunk 6]
```diff
@@ -1021,18 +1030,19 @@ def parse_quotable_robust(&block)
...
-            emit_row(row, &block)
+            emit_row(row, quoted_fields, &block)
             row = []
+            quoted_fields = []
           end
           skip_needless_lines
           start_row
         elsif @scanner.eos?
           break if row.empty? and value.nil?
           row << value
```
Comment (kou) [inline lib/csv/parser.rb:1044]: Do we need add an element to `quoted_field`s here?

```suggestion
          row << value
          quoted_fields << @quoted_column_value
```

File: lib/csv/writer.rb [hunk 7]
```diff
@@ -39,7 +39,7 @@ def <<(row)
       @headers ||= row if @use_headers
       @lineno += 1
 
-      row = @fields_converter.convert(row, nil, lineno) if @fields_converter
+      row = @fields_converter.convert(row, nil, lineno, []) if @fields_converter
```
Comment (kou) [inline lib/csv/writer.rb:42]:
```suggestion
      if @fields_converter
        quoted_fields = [false] * row.size
        row = @fields_converter.convert(row, nil, lineno, quoted_fields)
      end
```

File: lib/csv/writer.rb [hunk 8]
```diff
@@ -94,7 +94,7 @@ def prepare_header
       return unless @headers
 
       converter = @options[:header_fields_converter]
-      @headers = converter.convert(@headers, nil, 0)
+      @headers = converter.convert(@headers, nil, 0, [])
```
Comment (kou) [inline lib/csv/writer.rb:97]: We need to care about a case that `headers` is the following `String`:

```csv
"quoted",unquoted,"quoted-again"
```

Could you also add a test for this case?

File: test/csv/parse/test_convert.rb [hunk 9]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      @preserving_converter = lambda do |field, info|
+        f = field.encode(CSV::ConverterEncoding)
+        return field if info.quoted?
+        return DateTime.parse(f) if f.match?(DateTimeMatcher) rescue
+        return Integer(f) rescue
+        return Float(f) rescue
+        field
+      end
+
+      @str = <<~CSV
```
Comment (kou) [inline test/csv/parse/test_convert.rb:122]:
```suggestion
      @csv = <<~CSV
```

File: test/csv/parse/test_convert.rb [hunk 10]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
     assert_equal([nil, "empty", "a"],
                  CSV.parse_line(',"",a', empty_value: "empty"))
   end
+
+  sub_test_case("add quoted info") do
```
Comment (kou) [inline test/csv/parse/test_convert.rb:111]:
```suggestion
  sub_test_case("#quoted?") do
```

File: test/csv/parse/test_convert.rb [hunk 11]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+    def setup
+      @preserving_converter = lambda do |field, info|
+        f = field.encode(CSV::ConverterEncoding)
+        return field if info.quoted?
+        return DateTime.parse(f) if f.match?(DateTimeMatcher) rescue
+        return Integer(f) rescue
+        return Float(f) rescue
+        field
```
Comment (kou) [inline test/csv/parse/test_convert.rb:119]: Can we simplify this?

```suggestion
        f = field.encode(CSV::ConverterEncoding)
        return f if info.quoted?
        Integer(f, 10)
```

File: test/csv/parse/test_convert.rb [hunk 12]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      end
+
+      @str = <<~CSV
+        "serial",value
+        "109",12
+        "10A",13
+      CSV
+    end
+
+    def test_custom_field_info_quoted_parse_line
```
Comment (kou) [inline test/csv/parse/test_convert.rb:129]:
```suggestion
    def test_parse_line
```

File: test/csv/parse/test_convert.rb [hunk 13]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      row = CSV.parse_line('1,"2",3', converters: @preserving_converter)
+      assert_equal([1, "2", 3], row)
+    end
+
+    def test_custom_field_info_quoted_parse
```
Comment (kou) [inline test/csv/parse/test_convert.rb:134]:
```suggestion
    def test_parse
```

File: test/csv/parse/test_convert.rb [hunk 14]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      expected = [["serial", "value"], ["109", 12], ["10A", 13]]
+      row = CSV.parse(@str, converters: @preserving_converter)
```
Comment (kou) [inline test/csv/parse/test_convert.rb:136]: How about putting the parsed CSV (`@str`) here instead of creating it in `setup`?
It'll be easy to read what is tested here by aligning the expected value with input value.

File: test/csv/parse/test_convert.rb [hunk 15]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      assert_equal(expected, row)
+    end
+
+    def test_custom_field_info_quoted_parse_header
```
Comment (kou) [inline test/csv/parse/test_convert.rb:140]:
```suggestion
    def test_parse_headers
```

File: test/csv/parse/test_convert.rb [hunk 16]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+      quoted_header_converter = lambda do |field, info|
+        f = field.encode(CSV::ConverterEncoding)
+        return field if info.quoted?
+        field.to_sym
```
Comment (kou) [inline test/csv/parse/test_convert.rb:144]:
```suggestion
        f = field.encode(CSV::ConverterEncoding)
        return f if info.quoted?
        f.to_sym
```

File: test/csv/parse/test_convert.rb [hunk 17]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+
+    def test_custom_field_info_quoted_parse_line
+      row = CSV.parse_line('1,"2",3', converters: @preserving_converter)
+      assert_equal([1, "2", 3], row)
+    end
+
+    def test_custom_field_info_quoted_parse
+      expected = [["serial", "value"], ["109", 12], ["10A", 13]]
+      row = CSV.parse(@str, converters: @preserving_converter)
+      assert_equal(expected, row)
```
Comment (kou) [inline test/csv/parse/test_convert.rb:137]:
```suggestion
      rows = CSV.parse(@str, converters: @preserving_converter)
      assert_equal(expected, rows)
```

File: test/csv/parse/test_convert.rb [hunk 18]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+
+    def test_custom_field_info_quoted_parse_header
+      quoted_header_converter = lambda do |field, info|
+        f = field.encode(CSV::ConverterEncoding)
+        return field if info.quoted?
+        field.to_sym
+      end
+      expected = [["serial", :value], ["109", "12"], ["10A", "13"]]
+      row = CSV.parse(@str, headers: true, header_converters: quoted_header_converter)
+      assert_equal(expected, row.to_a)
```
Comment (kou) [inline test/csv/parse/test_convert.rb:148]:
```suggestion
      table = CSV.parse(@str, headers: true, header_converters: quoted_header_converter)
      assert_equal(expected, table.to_a)
```

File: test/csv/parse/test_convert.rb [hunk 19]
```diff
@@ -107,4 +107,45 @@ def test_empty_value
...
+    end
+
+    def test_custom_field_info_quoted_parse_header
+      quoted_header_converter = lambda do |field, info|
+        f = field.encode(CSV::ConverterEncoding)
+        return field if info.quoted?
+        field.to_sym
+      end
+      expected = [["serial", :value], ["109", "12"], ["10A", "13"]]
+      row = CSV.parse(@str, headers: true, header_converters: quoted_header_converter)
```
Comment (kou) [inline test/csv/parse/test_convert.rb:147]: Could you also add `CSV.parse(..., headers: '"serial",value', ...)` (specifying headers as `String`) case?

**Review by kou — COMMENTED** [review]
Thanks!
(If you want me to write code instead of commenting, please let me know.)

**Comment by heronshoes** [comment]
Thanks! I will check.
(I think this style is better because I can check one by one.)

**Comment by heronshoes** [comment]
I will add test case in:
- "quoted",unquoted,"quoted-again"
- Put expected @str in assertion locally
- CSV.parse(..., headers: '"serial",value', ...) (specifying headers as String) case

Please hold approving.

**Comment by heronshoes** [comment]
I added and fixed except 3rd one:
- CSV.parse(..., headers: '"serial",value', ...) (specifying headers as String) case

This one requires fix in `prepare_header`, `parse_headers` of parser.rb, but I can't fix yet.
It may also require `Symbol#encode` fix.

Current CSV has a bug in converting when specified header has a Symbol.
```ruby
CSV.parse("1,2", headers: ["string", :symbol], header_converters: :symbol)
```

`CSV.parse(..., headers: '"serial",value', ...)` case requires prior parsing, so same issue will happen.

**Comment by kou** [comment]
How about this?

```diff
diff --git a/lib/csv/parser.rb b/lib/csv/parser.rb
index c3c1eff..cd5a45f 100644
--- a/lib/csv/parser.rb
+++ b/lib/csv/parser.rb
@@ -758,9 +758,10 @@ class CSV
       case headers
       when Array
         @raw_headers = headers
+        quoted_fields = [false] * @raw_headers.size
         @use_headers = true
       when String
-        @raw_headers = parse_headers(headers)
+        @raw_headers, quoted_fields = parse_headers(headers)
         @use_headers = true
       when nil, false
         @raw_headers = nil
@@ -770,20 +771,24 @@ class CSV
         @use_headers = true
       end
       if @raw_headers
-        # Set 'quoted?' info to false to avoid re-converting
-        @headers = adjust_headers(@raw_headers, [true] * @raw_headers.size)
+        @headers = adjust_headers(@raw_headers, quoted_fields)
       else
         @headers = nil
       end
     end
 
     def parse_headers(row)
-      converters = @return_headers ? [] : @header_fields_converter.to_a
-      CSV.parse_line(row,
-                     col_sep:    @column_separator,
-                     row_sep:    @row_separator,
-                     quote_char: @quote_character,
-                     converters: converters)
+      quoted_fields = []
+      converter = lambda do |field, info|
+        quoted_fields << info.quoted?
+        field
+      end
+      headers = CSV.parse_line(row,
+                               col_sep:    @column_separator,
+                               row_sep:    @row_separator,
+                               quote_char: @quote_character,
+                               converters: [converter])
+      [headers, quoted_fields]
     end
 
     def adjust_headers(headers, quoted_fields)
```

**Comment by heronshoes** [comment]
Thank you!! I reflected this code!

File: lib/csv/parser.rb [hunk 20]
```diff
@@ -770,21 +771,28 @@ def prepare_header
...
+      quoted_fields = []
+      converter = lambda do |field, info|
+        quoted_fields << info.quoted?
+        field
+      end
+      headers = CSV.parse_line( row,
+                                col_sep:    @column_separator,
+                                row_sep:    @row_separator,
+                                quote_char: @quote_character,
+                                converters: [converter])
```
Comment (kou) [inline lib/csv/parser.rb:790]:
```suggestion
      headers = CSV.parse_line(row,
                               col_sep:    @column_separator,
                               row_sep:    @row_separator,
                               quote_char: @quote_character,
                               converters: [converter])
```

File: lib/csv/parser.rb [hunk 21]
```diff
@@ -928,9 +936,11 @@ def parse_no_quote(&block)
         if line.empty?
           next if @skip_blanks
           row = []
+          quoted_fields = []
         else
           line = strip_value(line)
           row = line.split(@split_column_separator, -1)
+          quoted_fields = []
```
Comment (kou) [inline lib/csv/parser.rb:943]:
```suggestion
          quoted_fields = [false] * row.size
```

File: test/csv/parse/test_convert.rb [hunk 22]
```diff
@@ -107,4 +107,69 @@ def test_empty_value
...
+        "109",1
+        "10A",2
+      CSV
+      assert_equal(expected, table.to_a)
+    end
+  end
+
+  sub_test_case("converter with #encode for Symbols") do
+    def test_encode_symbol
+      assert_nothing_raised { CSV.parse("1,2", headers: ["string", :symbol] , header_converters: :symbol) }
```
Comment (kou) [inline test/csv/parse/test_convert.rb:172]: It seems that this is not related to this pull request.
Could you remove this from this pull request? If this is needed, please open another pull request.

**Review by kou — COMMENTED** [review]
Thanks!
It looks almost good! Could you confirm my comments?

File: test/csv/parse/test_convert.rb (hunk 22 above)
Comment (heronshoes) [reply test/csv/parse/test_convert.rb]: Thank you. I removed last one and create modified commit.

I will consider more for another pull request about converting when specified header has a Symbol.

**Comment by kou** [comment]
Thanks!
